pkg/metrics/prometheus.go

```diff
@@ -0,0 +1,183 @@
+	p.reg.MustRegister(
+		p.decodeDuration,
+		p.encodeDuration,
+		p.cropDuration,
+		p.scaleDuration,
+		p.resizeDuration,
+		p.grayscaleDuration,
+		p.blurDuration,
+		p.fixOrientationDuration,
+		p.flipDuration,
+		p.rotateDuration,
+		p.storageGetErrors,
+		p.processorErrors,
+	)
+}
+
+func (p prometheusService) TrackDecodeDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.decodeDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackEncodeDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.encodeDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackCropDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.cropDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackScaleDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.scaleDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackResizeDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.resizeDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackGrayScaleDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.grayscaleDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackBlurDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.blurDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackFixOrientationDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.fixOrientationDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackFlipDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.flipDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) TrackRotateDuration(start time.Time, ImageData []byte) {
+	labelValue := p.getLabelValue(ImageData)
+	p.rotateDuration.WithLabelValues(labelValue).Observe(time.Since(start).Seconds())
+}
+
+func (p prometheusService) CountStorageGetErrors() {
+	p.storageGetErrors.Inc()
+}
+
+func (p prometheusService) CountProcessorErrors() {
+	p.processorErrors.Inc()
+}
+
+func (p prometheusService) getLabelValue(ImageData []byte) string {
+	ext := strings.Split(http.DetectContentType(ImageData), "/")[1]
+	labelValue := fmt.Sprintf("%s.%s", GetImageSizeCluster(ImageData), ext)
+	return labelValue
+}
+
+func (p prometheusService) AddMetricsEndPoint(metricsPath string, router *mux.Router) {
```

Comment:
See if this can be set during object creation and no method is required.

Reply:
Then object creation, `NewPrometheus()` needs to be called at routes.go and we can't set the `MetricService ` for Manipulator which is initialised in `NewDependencies()`